spec/ready_set_spec.rb

```diff
@@ -5,6 +5,114 @@
     expect(Readyset::VERSION).not_to be nil
   end
 
+  describe '.create_cache!' do
+    let(:query) { build(:seen_but_not_cached_query) }
+
+    context 'when given neither a SQL string nor an ID' do
+      subject { Readyset.create_cache! }
+
+      it 'raises an ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
```

Comment:
Ooo interesting! I was thinking that using `subject` was part of the "arrange" step of the specs, since it makes it obvious exactly what's under test. It also DRYs up the tests a bit because if the subject changes at all (e.g. the name of the method changes or it takes a new parameter), we only have to make the change in one place (in the `subject`). It also facilitates the use of shared examples, since the shared examples can just reference `subject` in order to run the test rather than having to pass additional context to the shared examples via a `let`.

How do you think `subject` impacts readability?

Reply:
**TLDR:** `subject` performs `.create_cache`, the main action.  

I think the specs are solid and mergeable, as-is, but it's fun to talk about testing in the abstract.

`subject`, in this case, contains the main action whose output we're interested in, so it would fall under *act* or *exercise*. *Arrange* consists of setting up the environment/context, such as mocks, around the main action, `.create_cache`.

Individual specs can be thought of as NO-PARKING signs. If there are conditions, they should be understandable at a glance as you're driving by (otherwise you end up like [Los Angeles](https://lamag.com/.image/c_limit%2Ccs_srgb%2Cq_auto:eco%2Cw_660/MTk3NTU2NDE3MzY1MDI2NDk2/image-placeholder-title.webp)). 

> "readable at a glance" > "DRY code"

```ruby
context 'when given both a SQL string and an ID' do
  it 'raises an ArgumentError' do
    # Setup
    query = build(:seen_but_not_cached)
    sql_string = query.text
    query_id = query.id

    # Exercise & Verify
    with_both_args = Readyset.create_cache!(sql: sql_string, id: query_id)
    
    expect { with_both_args }.to raise_error(ArgumentError)
  end
end
```
So, readability in the original spec was solid enough IMO. But a less-DRY rewrite explicitly details `:seen_but_not_cached` as part of the *arrange/setup* step. One could find that detail, but they would have to hunt for it. Also, the query `let` is re-defined later on, which introduces some obscurity in verifying what we are testing.

In regards to `shared_examples`, I think using them is better than not using them. But sometimes they also over-abstract tests, hiding away logic or behavior. I have mixed feelings about them. 

---

If you think this code will survive into v2, then I think being pedantic keeps code from aging faster (all code ages like milk, we determine how fast). Though, I know we're working with a time-crunch. The cost/benefit leans in favor of at least removing the `let`s blocks, but everything is solid enough.

Reply:
You have convinced me! At a previous role of mine, we made heavy use of `subject`, `before`, `let`, etc. which is why my specs all look the way they do now. I think in my mind, they separated (to some extent) the "setup", "exercise" and "verify" steps of testing -- `before` is "setup", `subject` is "exercise", `it` is "verify".

But I certainly see the value in having the examples themselves contain all the phases of the test -- all of the context for the test is in one place. I also just found thoughtbot's rspec style guide, which has a lot of great pointers! https://github.com/thoughtbot/guides/tree/main/testing-rspec